### PR TITLE
SW-5990 Some questionnaire deliverables need to be saved 2x to be marked as complete

### PR DESCRIPTION
**Pull request by constanzauanini**

### CODE REVIEW

File: src/utils/documentProducer/variables.ts [hunk 1]
```diff
@@ -81,16 +82,22 @@ export const getRawValue = (variable: VariableWithValues): number | number[] | s
...
   const allRequiredVariables = variablesWithValues.filter(
     (v) => v.isRequired && variableDependencyMet(v, variablesWithValues)
   );
-
   const missingRequiredFields = allRequiredVariables.some((variable) => {
     let hasEmptyValue = false;
     const values = variable.values;
     if (!values || values.length === 0 || getRawValue(variable) === undefined || getRawValue(variable) === '') {
       hasEmptyValue = true;
+      if (pendingTableValues?.get(variable.id)?.length) {
```
Comment (tommylau523) [inline src/utils/documentProducer/variables.ts:97]: This may be a lot to add, but we may need to do the following:
1) Add isRequired to the `VariableTableCell` type, this can be passed down from `TableVariableWithValues` in `getInitialCellValues`.
2) Check if an entire row is empty. If so, we can treat it like the row does not exist
3) For a row, check if a column is required. If it is, check if the value is not empty
4) Then we need to check a required table has at least one row (which we have here!)

Comment (tommylau523) [reply src/utils/documentProducer/variables.ts:127]: If just one cell is needed, we do need to check if at least one cell is non-empty. Basically still step 2 and 4. A user right now can type something, hit backspace and submit an empty string.

**Review by tommylau523 — APPROVED** [review]
Looks like this is working correctly. Ok with submitting for now till we figure out what a required table means!

File: src/hooks/useProjectVariablesUpdate/index.ts [hunk 2]
```diff
@@ -96,7 +96,7 @@ export const useProjectVariablesUpdate = (
   }, [pendingVariableValues, variablesWithValues]);
 
   const missingFields = useMemo(
-    () => missingRequiredFields(stagedVariableWithValues),
+    () => missingRequiredFields(stagedVariableWithValues, pendingCellValues),
     [missingRequiredFields, stagedVariableWithValues]
```
Comment (tommylau523) [inline src/hooks/useProjectVariablesUpdate/index.ts:100]: need to add pendingCellValues to deps

**Comment by constanzauanini** [comment]
> Looks like this is working correctly. Ok with submitting for now till we figure out what a required table means!

Just added a small change to check for empty values

**Comment by tommylau523** [comment]
Thanks Constanza! One comment about adding a deps to useCallback

**Review by tommylau523 — CHANGES_REQUESTED** [review]
I think we need to update the cell checking

File: src/hooks/useProjectVariablesUpdate/index.ts (hunk 2 above)
Comment (constanzauanini) [reply src/hooks/useProjectVariablesUpdate/index.ts:100]: Thanks!

File: src/utils/documentProducer/variables.ts [hunk 3]
```diff
@@ -112,15 +113,24 @@ export const isValueEmpty = (
...
+export const missingRequiredFields = (
+  variablesWithValues: VariableWithValues[],
+  pendingTableValues?: Map<number, VariableTableCell[][]>
+) => {
   const allRequiredVariables = variablesWithValues.filter(
     (v) => v.isRequired && variableDependencyMet(v, variablesWithValues)
   );
-
   const missingRequiredFields = allRequiredVariables.some((variable) => {
     const values = variable.values;
```
Comment (tommylau523) [inline src/utils/documentProducer/variables.ts:124]: I think this may work and we can move it to the top of the function. 

```
if (variable.type === 'Table') {
  const tableCells = !pendingTableValues?.get(variable.id)

  // Return true if there is no row
  if (!tableCells || tableCells.length === 0) {
    return true;
  }
  
  // Return true if every cell is empty
  return tableCells.every((row) => row.every(cell) => {
    if (!cell.values || cell.values.length ===0) {
        return true;
    } 
    return cellValue(cell.values[0]).toString() === ''
  }))
}
```

**Review by tommylau523 — APPROVED** [review]

File: src/utils/documentProducer/variables.ts (hunk 3 above)
Comment (tommylau523) [reply src/utils/documentProducer/variables.ts:124]: Ahh I understand.  Quite tricky because I think we need to catch the case of not adding a row to an empty table, as well as not modifying an existing table.

Comment (tommylau523) [reply src/utils/documentProducer/variables.ts:124]: Perhaps we can add a `stagedCellValues` similar to `stagedVariableValues`

Comment (constanzauanini) [reply src/utils/documentProducer/variables.ts:124]: Yes! That was my first idea but was getting trouble with typescript

Comment (constanzauanini) [reply src/utils/documentProducer/variables.ts:124]: But if there is no row, then values of the table will be empty, so it will be marked as required in the first check

**Comment by tommylau523** [comment]
I did test in Vercel, and submitting nothing in a table will still complete the Deliverable.

**Comment by constanzauanini** [comment]
> I did test in Vercel, and submitting nothing in a table will still complete the Deliverable.

Oh, thanks for testing @tommylau523. I was trying to see what was the values of "pendingValues" when saving an empty table. 

Do you want to create a PR with the code you did to add the pendingValues to the stages variables?

**Comment by tommylau523** [comment]
> > I did test in Vercel, and submitting nothing in a table will still complete the Deliverable.
> 
> Oh, thanks for testing @tommylau523. I was trying to see what was the values of "pendingValues" when saving an empty table. 
> 
> Do you want to create a PR with the code you did to add the pendingValues to the stages variables?

I am trying a few more things for now! I haven't really tested it out yet. Sorry for the back and forth, I was also just thinking of all the edge cases. I will open a PR if it ends up working!

**Comment by constanzauanini** [comment]
> > > I did test in Vercel, and submitting nothing in a table will still complete the Deliverable.
> > 
> > 
> > Oh, thanks for testing @tommylau523. I was trying to see what was the values of "pendingValues" when saving an empty table.
> > Do you want to create a PR with the code you did to add the pendingValues to the stages variables?
> 
> I am trying a few more things for now! I haven't really tested it out yet. Sorry for the back and forth, I was also just thinking of all the edge cases. I will open a PR if it ends up working!

Ok, just checked that pendingValues has the table headers always, that's why it was failing.

**Comment by tommylau523** [comment]
https://github.com/terraware/terraware-web/pull/3166 

This seems to work for me!
